File: surface/widgets/compose/components/widgets/feedback/Alert.kt
// ============================================================
// Clef Surface Compose Widget — Alert
//
// Inline, persistent status message that communicates important
// information within the layout. Unlike a toast, an alert does
// not auto-dismiss -- it remains visible until the user
// explicitly closes it (when closable) or until the triggering
// condition is resolved.
//
// Compose adaptation: Card with variant-colored leading icon,
// title, optional description, and optional close IconButton.
// See widget spec: repertoire/widgets/feedback/alert.widget
// ============================================================

package clef.surface.compose.components.widgets.feedback

import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.Alignment
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.unit.dp
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Info
import androidx.compose.material.icons.filled.Warning
import androidx.compose.material.icons.filled.Error
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.Close

// --------------- Variant Configuration ---------------

enum class AlertVariant(
    val icon: ImageVector,
    val tint: Color,
    val containerColor: Color,
) {
    Info(
        icon = Icons.Filled.Info,
        tint = Color(0xFF1976D2),
        containerColor = Color(0xFFE3F2FD),
    ),
    Warning(
        icon = Icons.Filled.Warning,
        tint = Color(0xFFF57C00),
        containerColor = Color(0xFFFFF3E0),
    ),
    Error(
        icon = Icons.Filled.Error,
        tint = Color(0xFFD32F2F),
        containerColor = Color(0xFFFFEBEE),
    ),
    Success(
        icon = Icons.Filled.CheckCircle,
        tint = Color(0xFF388E3C),
        containerColor = Color(0xFFE8F5E9),
    ),
}

// --------------- Component ---------------

/**
 * Inline, persistent status message with icon, title, optional
 * description, and an optional close button.
 *
 * @param variant Visual variant controlling icon, colors, and semantics.
 * @param title Primary alert message.
 * @param description Optional secondary detail or guidance text.
 * @param closable Whether the alert can be dismissed by the user.
 * @param onClose Callback fired when the alert is dismissed.
 * @param modifier Modifier applied to the root Card.
 * @param content Additional composable content rendered inside the alert body.
 */
@Composable
fun Alert(
    variant: AlertVariant = AlertVariant.Info,
    title: String? = null,
    description: String? = null,
    closable: Boolean = false,
    onClose: (() -> Unit)? = null,
    modifier: Modifier = Modifier,
    content: @Composable (ColumnScope.() -> Unit)? = null,
) {
    var dismissed by remember { mutableStateOf(false) }

    if (dismissed) return

    Card(
        modifier = modifier.fillMaxWidth(),
        colors = CardDefaults.cardColors(
            containerColor = variant.containerColor,
        ),
    ) {
        Column(
            modifier = Modifier.padding(16.dp),
        ) {
            // Title row with icon and optional close button
            Row(
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier.fillMaxWidth(),
            ) {
                Icon(
                    imageVector = variant.icon,
                    contentDescription = variant.name,
                    tint = variant.tint,
                    modifier = Modifier.size(20.dp),
                )

                Spacer(modifier = Modifier.width(12.dp))

                if (title != null) {
                    Text(
                        text = title,
                        style = MaterialTheme.typography.titleSmall,
                        color = variant.tint,
                        modifier = Modifier.weight(1f),
                    )
                } else {
                    Spacer(modifier = Modifier.weight(1f))
                }

                if (closable) {
                    IconButton(
                        onClick = {
                            dismissed = true
                            onClose?.invoke()
                        },
                        modifier = Modifier.size(24.dp),
                    ) {
                        Icon(
                            imageVector = Icons.Filled.Close,
                            contentDescription = "Dismiss",
                            modifier = Modifier.size(16.dp),
                        )
                    }
                }
            }

            // Description
            if (description != null) {
                Spacer(modifier = Modifier.height(4.dp))
                Text(
                    text = description,
                    style = MaterialTheme.typography.bodySmall,
                    modifier = Modifier.padding(start = 32.dp),
                )
            }

            // Additional children
            if (content != null) {
                Spacer(modifier = Modifier.height(8.dp))
                Column(
                    modifier = Modifier.padding(start = 32.dp),
                    content = content,
                )
            }
        }
    }
}

File: surface/widgets/swiftui/components/widgets/feedback/Alert.swift
// ============================================================
// Clef Surface SwiftUI Widget — Alert
//
// Inline, persistent status message that communicates important
// information within the layout. Supports info, warning, error,
// and success variants with optional close button.
//
// Adapts the alert.widget spec to SwiftUI rendering.
// ============================================================

import SwiftUI

// --------------- Types ---------------

enum AlertVariant {
    case info, warning, error, success

    var iconName: String {
        switch self {
        case .info: return "info.circle.fill"
        case .warning: return "exclamationmark.triangle.fill"
        case .error: return "exclamationmark.circle.fill"
        case .success: return "checkmark.circle.fill"
        }
    }

    var tintColor: Color {
        switch self {
        case .info: return Color(red: 0.1, green: 0.46, blue: 0.82)
        case .warning: return Color(red: 0.96, green: 0.49, blue: 0)
        case .error: return Color(red: 0.83, green: 0.18, blue: 0.18)
        case .success: return Color(red: 0.22, green: 0.56, blue: 0.24)
        }
    }

    var containerColor: Color {
        switch self {
        case .info: return Color(red: 0.89, green: 0.95, blue: 0.99)
        case .warning: return Color(red: 1.0, green: 0.95, blue: 0.88)
        case .error: return Color(red: 1.0, green: 0.92, blue: 0.93)
        case .success: return Color(red: 0.91, green: 0.96, blue: 0.91)
        }
    }
}

// --------------- Component ---------------

/// Alert view for inline, persistent status messages.
///
/// - Parameters:
///   - variant: Visual variant controlling icon, colors, and semantics.
///   - title: Primary alert message.
///   - description: Optional secondary detail text.
///   - closable: Whether the alert can be dismissed.
///   - onClose: Callback fired when the alert is dismissed.
///   - content: Additional content rendered inside the alert body.
struct AlertView<Content: View>: View {
    var variant: AlertVariant = .info
    var title: String? = nil
    var description: String? = nil
    var closable: Bool = false
    var onClose: (() -> Void)? = nil
    @ViewBuilder var content: Content

    @State private var dismissed = false

    var body: some View {
        if !dismissed {
            VStack(alignment: .leading, spacing: 4) {
                HStack(alignment: .center, spacing: 12) {
                    Image(systemName: variant.iconName)
                        .foregroundColor(variant.tintColor)
                        .frame(width: 20, height: 20)

                    if let title = title {
                        Text(title)
                            .font(.subheadline)
                            .fontWeight(.semibold)
                            .foregroundColor(variant.tintColor)
                            .frame(maxWidth: .infinity, alignment: .leading)
                    } else {
                        Spacer()
                    }

                    if closable {
                        SwiftUI.Button(action: {
                            dismissed = true
                            onClose?()
                        }) {
                            Image(systemName: "xmark")
                                .font(.caption)
                                .foregroundColor(.secondary)
                        }
                        .buttonStyle(.plain)
                        .accessibilityLabel("Dismiss")
                    }
                }

                if let description = description {
                    Text(description)
                        .font(.caption)
                        .foregroundColor(.primary)
                        .padding(.leading, 32)
                }

                content
                    .padding(.leading, 32)
            }
            .padding(16)
            .background(variant.containerColor)
            .clipShape(RoundedRectangle(cornerRadius: 8))
        }
    }
}

extension AlertView where Content == EmptyView {
    init(
        variant: AlertVariant = .info,
        title: String? = nil,
        description: String? = nil,
        closable: Bool = false,
        onClose: (() -> Void)? = nil
    ) {
        self.variant = variant
        self.title = title
        self.description = description
        self.closable = closable
        self.onClose = onClose
        self.content = { EmptyView() }
    }
}
